--- Linux-PAM-1.1.4/doc/Makefile.am
+++ Linux-PAM-1.1.4/doc/Makefile.am
@@ -2,7 +2,7 @@
 # Copyright (c) 2005, 2006 Thorsten Kukuk <kukuk@suse.de>
 #
 
-SUBDIRS = man specs sag adg mwg
+SUBDIRS = man sag adg mwg
 
 CLEANFILES = *~
 
--- Linux-PAM-1.1.4/Makefile.am
+++ Linux-PAM-1.1.4/Makefile.am
@@ -5,9 +5,9 @@
 AUTOMAKE_OPTIONS = 1.9 gnu dist-bzip2 check-news
 
 if STATIC_MODULES
-SUBDIRS = modules libpam libpamc libpam_misc tests po conf doc examples xtests
+SUBDIRS = modules libpam libpamc libpam_misc doc
 else
-SUBDIRS = libpam tests libpamc libpam_misc modules po conf doc examples xtests
+SUBDIRS = libpam libpamc libpam_misc modules doc
 endif
 
 CLEANFILES = *~
