--- e2fsprogs-1.41.14/Makefile.in
+++ e2fsprogs-1.41.14/Makefile.in
@@ -15,8 +15,8 @@
 @BLKID_CMT@BLKID_LIB_SUBDIR= lib/blkid
 
 LIB_SUBDIRS=lib/et lib/ss lib/e2p $(UUID_LIB_SUBDIR) lib/ext2fs $(BLKID_LIB_SUBDIR) intl
-PROG_SUBDIRS=e2fsck $(DEBUGFS_DIR) misc $(RESIZE_DIR) tests/progs po
-SUBDIRS=util $(LIB_SUBDIRS) $(PROG_SUBDIRS) tests
+PROG_SUBDIRS=e2fsck $(DEBUGFS_DIR) misc $(RESIZE_DIR)
+SUBDIRS=util $(LIB_SUBDIRS) $(PROG_SUBDIRS)
 
 SUBS= lib/ext2fs/ext2_types.h lib/blkid/blkid_types.h lib/uuid/uuid_types.h
 
@@ -25,7 +25,6 @@
 all:: subs
 	$(MAKE) libs
 	$(MAKE) progs
-	$(MAKE) docs
 
 subs: $(DEP_SUBSTITUTE)
 	@for i in $(SUBS) ; do if test -d `dirname $$i` ; \
@@ -59,11 +58,11 @@
 	-test -d doc && cd doc && $(MAKE) distclean
 
 install: subs all-libs-recursive install-progs-recursive \
-  install-shlibs-libs-recursive install-doc-libs
+  install-shlibs-libs-recursive 
 	if test ! -d e2fsck && test ! -d debugfs && test ! -d misc && test ! -d ext2ed ; then $(MAKE) install-libs ; fi
 
 install-strip: subs all-libs-recursive install-strip-progs-recursive \
-  install-shlibs-strip-libs-recursive install-doc-libs
+  install-shlibs-strip-libs-recursive
 
 uninstall: uninstall-progs-recursive uninstall-shlibs-libs-recursive uninstall-doc-libs
 
