diff -up ./mozilla/security/nss/cmd/fipstest/fipstest.c.801975 ./mozilla/security/nss/cmd/fipstest/fipstest.c
--- ./mozilla/security/nss/cmd/fipstest/fipstest.c.801975	2011-09-17 14:10:05.000000000 -0700
+++ ./mozilla/security/nss/cmd/fipstest/fipstest.c	2012-04-10 11:14:55.700732586 -0700
@@ -4940,6 +4940,9 @@ loser:
 int main(int argc, char **argv)
 {
     if (argc < 2) exit (-1);
+	
+    RNG_RNGInit();
+    SECOID_Init();
 
     /*************/
     /*   TDEA    */
