--- shadow-4.0.4.1.orig/src/usermod.c	2004-09-27 17:35:47.911752832 +0000
+++ shadow-4.0.4.1/src/usermod.c	2004-09-27 17:30:38.440799568 +0000
@@ -1148,6 +1148,8 @@ static void usr_update (void)
 	}
 #ifdef	SHADOWPWD
 	if (spwd && (lflg || eflg || fflg || pflg || Lflg || Uflg)) {
+		if (pflg)
+			spent.sp_lstchg = time ((time_t *) 0) / SCALE;
 		if (!spw_update (&spent)) {
 			fprintf (stderr,
 				 _
