diff -uNrp shadow-4.0.4.1/src/usermod.c shadow-4.0.4.1.owl-usermod-unlock/src/usermod.c
--- shadow-4.0.4.1/src/usermod.c	2003-06-30 13:17:51 +0000
+++ shadow-4.0.4.1.owl-usermod-unlock/src/usermod.c	2004-02-28 22:58:52 +0000
@@ -330,7 +330,7 @@ static char *new_pw_passwd (char *pw_pas
 		strcpy (buf, "!");
 		strcat (buf, pw_pass);
 		pw_pass = buf;
-	} else if (Uflg && pw_pass[0] == '!') {
+	} else if (Uflg && pw_pass[0] == '!' && pw_pass[1] != '!') {
 		char *s;
 
 		SYSLOG ((LOG_INFO, "unlock user `%s' password", pw_name));
