diff -uNrp shadow-4.0.4.1/man/getspnam.3 shadow-4.0.4.1.owl-man/man/getspnam.3
--- shadow-4.0.4.1/man/getspnam.3	1970-01-01 00:00:00 +0000
+++ shadow-4.0.4.1.owl-man/man/getspnam.3	2004-02-28 23:13:02 +0000
@@ -0,0 +1 @@
+.so shadow.3
diff -uNrp shadow-4.0.4.1/man/Makefile.am shadow-4.0.4.1.owl-man/man/Makefile.am
--- shadow-4.0.4.1/man/Makefile.am	2003-12-17 10:04:25 +0000
+++ shadow-4.0.4.1.owl-man/man/Makefile.am	2004-02-28 23:13:02 +0000
@@ -11,6 +11,7 @@ man_MANS = \
 	expiry.1 \
 	faillog.5 \
 	faillog.8 \
+	getspnam.3 \
 	gpasswd.1 \
 	groupadd.8 \
 	groupdel.8 \
