diff -uNp -r shadow-4.0.4.1.orig/lib/getdef.c shadow-4.0.4.1/lib/getdef.c
--- shadow-4.0.4.1.orig/lib/getdef.c	2011-01-20 09:25:33 +0000
+++ shadow-4.0.4.1/lib/getdef.c	2011-01-20 09:27:08 +0000
@@ -82,6 +82,7 @@ static struct itemdef def_table[] = {
 	{ "GID_MAX",			NULL },
 	{ "GID_MIN",			NULL },
 	{ "GROUPNAME_MAX",		NULL },
+	{ "GROUPNAME_RELAXED",		NULL },
 	{ "HUSHLOGIN_FILE",		NULL },
 	{ "ISSUE_FILE",			NULL },
 	{ "KILLCHAR",			NULL },
@@ -128,6 +129,7 @@ static struct itemdef def_table[] = {
 	{ "USERDEL_CMD",		NULL },
 	{ "USERGROUPS_ENAB",		NULL },
 	{ "USERNAME_MAX",		NULL },
+	{ "USERNAME_RELAXED",		NULL },
 	{ "USE_TCB",			NULL }
 };
 
diff -uNp -r shadow-4.0.4.1.orig/libmisc/chkname.c shadow-4.0.4.1/libmisc/chkname.c
--- shadow-4.0.4.1.orig/libmisc/chkname.c	2011-01-20 09:25:33 +0000
+++ shadow-4.0.4.1/libmisc/chkname.c	2011-01-31 15:26:12 +0000
@@ -16,16 +16,19 @@ RCSID ("$Id: chkname.c,v 1.8 2003/05/12 
 #else
 #include <utmp.h>
 #endif
-static int good_name (const char *name)
+static int good_name (const char *name, int relaxed)
 {
 	/*
-	 * User/group names must match [a-z_][a-z0-9_-]*
+	 * User/group names must match [a-zA-Z_][a-zA-Z0-9_-]*
 	 */
-	if (!*name || !((*name >= 'a' && *name <= 'z') || *name == '_'))
+	if (!*name || !((*name >= 'a' && *name <= 'z') ||
+	                (*name >= 'A' && *name <= 'Z' && relaxed) ||
+	                 *name == '_'))
 		return 0;
 
 	while (*++name) {
 		if (!((*name >= 'a' && *name <= 'z') ||
+		      (*name >= 'A' && *name <= 'Z' && relaxed) ||
 		      (*name >= '0' && *name <= '9') ||
 		      *name == '_' || *name == '-' ||
 		      (*name == '$' && *(name + 1) == '\0')))
@@ -48,6 +51,7 @@ int check_user_name (const char *name)
 #else
 	struct utmp ut;
 #endif
+	int relaxed = getdef_bool("USERNAME_RELAXED");
 
 	/*
 	 * User names are limited by whatever utmp can
@@ -57,16 +61,18 @@ int check_user_name (const char *name)
 	if (strlen (name) > max_len)
 		return 0;
 
-	return good_name (name);
+	return good_name (name, relaxed);
 }
 
 int check_group_name (const char *name)
 {
+        int relaxed = getdef_bool("GROUPNAME_RELAXED");
+
 	/*
 	 * Arbitrary limit for group names.
 	 */
 	if (strlen (name) > min (getdef_unum ("GROUPNAME_MAX", 16), 32))
 		return 0;
 
-	return good_name (name);
+	return good_name (name, relaxed);
 }
