--- shadow-4.0.4.1/libmisc/pwdcheck.c.orig	Tue Apr 22 10:59:22 2003
+++ shadow-4.0.4.1/libmisc/pwdcheck.c	Fri Jun 11 05:50:42 2004
@@ -32,9 +32,9 @@ passwd_check (const char *user, const ch
 		goto bailout;
 
 	retcode = pam_acct_mgmt (pamh, 0);
-	if (retcode == PAM_NEW_AUTHTOK_REQD) {
+	if (retcode == PAM_NEW_AUTHTOK_REQD)
 		retcode = pam_chauthtok (pamh, PAM_CHANGE_EXPIRED_AUTHTOK);
-	} else if (retcode)
+	if (retcode)
 		goto bailout;
 
 	if (pam_setcred (pamh, 0))
