diff -puNr shadow-4.0.4.1.orig/libmisc/xmalloc.c shadow-4.0.4.1/libmisc/xmalloc.c
--- shadow-4.0.4.1.orig/libmisc/xmalloc.c	Tue Apr 22 10:59:22 2003
+++ shadow-4.0.4.1/libmisc/xmalloc.c	Fri Jan  7 02:30:51 2005
@@ -13,13 +13,13 @@
 RCSID ("$Id: xmalloc.c,v 1.4 2003/04/22 10:59:22 kloczek Exp $")
 #include <stdio.h>
+#include <stdlib.h>
 #include "defines.h"
-extern char *malloc ();
 
 char *xmalloc (size_t size)
 {
 	char *ptr;
 
-	ptr = malloc (size);
+	ptr = (char *)malloc (size);
 	if (!ptr && size) {
 		fprintf (stderr, _("malloc(%d) failed\n"), (int) size);
 		exit (13);
