Index: src/userdel.c
===================================================================
RCS file: /cvsroot/shadow/src/userdel.c,v
retrieving revision 1.29
retrieving revision 1.30
diff -u -p -r1.29 -r1.30
--- src/userdel.c	17 Dec 2003 09:43:30 -0000	1.29
+++ src/userdel.c	8 Oct 2004 04:01:32 -0000	1.30
@@ -30,7 +30,7 @@
 #include <config.h>
 
 #include "rcsid.h"
-RCSID (PKG_VER "$Id: userdel.c,v 1.29 2003/12/17 09:43:30 kloczek Exp $")
+RCSID (PKG_VER "$Id: userdel.c,v 1.30 2004/10/08 04:01:32 kloczek Exp $")
 #include <sys/stat.h>
 #include <stdio.h>
 #include <errno.h>
@@ -655,7 +655,7 @@ int main (int argc, char **argv)
 	if (retval != PAM_SUCCESS) {
 		fprintf (stderr,
 			 _("%s: PAM authentication failed\n"), Prog);
-		exit (1);
+		exit (E_PW_UPDATE);
 	}
 #endif				/* USE_PAM */
 
@@ -810,9 +810,10 @@ int main (int argc, char **argv)
 			pam_end (pamh, retval);
 	}
 
-	if (retval != PAM_SUCCESS)
+	if (retval != PAM_SUCCESS) {
 		fprintf (stderr, _("%s: PAM chauthtok failed\n"), Prog);
-	exit (1);
+		exit (E_PW_UPDATE);
+	}
 
 	if (retval == PAM_SUCCESS)
 		pam_end (pamh, PAM_SUCCESS);
