# DP: PR ld/12654: Pproperly handle R_386_TLS_LDO_32 for PIE.

--- a/bfd/elf32-i386.c
+++ b/bfd/elf32-i386.c
@@ -4022,7 +4022,7 @@
 	  break;
 
 	case R_386_TLS_LDO_32:
-	  if (info->shared || (input_section->flags & SEC_CODE) == 0)
+	  if (!info->executable || (input_section->flags & SEC_CODE) == 0)
 	    relocation -= elf_i386_dtpoff_base (info);
 	  else
 	    /* When converting LDO to LE, we must negate.  */
