--- browser/components/preferences/advanced.js	2009-11-24 22:12:34.000000000 -0600
+++ browser/components/preferences/advanced.js-new	2009-12-02 10:26:18.378458863 -0600
@@ -408,7 +408,7 @@
     var enabledPref = document.getElementById("app.update.enabled");
     var enableAppUpdate = document.getElementById("enableAppUpdate");
 
-    enableAppUpdate.disabled = !aus.canCheckForUpdates || enabledPref.locked;
+    enableAppUpdate.disabled = true;
   },
 
   /**
@@ -423,8 +423,7 @@
     var updateModeLabel = document.getElementById("updateModeLabel");
     var updateMode = document.getElementById("updateMode");
     
-    var disable = enabledPref.locked || !enabledPref.value ||
-                  autoPref.locked;
+    var disable = true;
     updateModeLabel.disabled = updateMode.disabled = disable;
   },
 
@@ -440,8 +439,7 @@
     
     var warnIncompatible = document.getElementById("warnIncompatible");
     
-    var disable = enabledPref.locked || !enabledPref.value || autoPref.locked ||
-                  !autoPref.value || modePref.locked;
+    var disable = true;
     warnIncompatible.disabled = disable;
   },
