diff -Naur pacman-4.0.0-/lib/libalpm/add.c pacman-4.0.0/lib/libalpm/add.c
--- pacman-4.0.0-/lib/libalpm/add.c	2011-10-12 12:33:02.000000000 +0000
+++ pacman-4.0.0/lib/libalpm/add.c	2011-10-30 02:30:24.708144434 +0000
@@ -653,6 +653,9 @@
 		char *scriptlet = _alpm_local_db_pkgpath(db, newpkg, "install");
 		const char *scriptlet_name = is_upgrade ? "post_upgrade" : "post_install";
 
+		if(is_upgrade) {
+			setenv("PKGNAME", oldpkg->name, 1);
+		}
 		_alpm_runscriptlet(handle, scriptlet, scriptlet_name,
 				newpkg->version, oldpkg ? oldpkg->version : NULL, 0);
 		free(scriptlet);
