From d2706a05513134f961e2500e0e596f2769707ce5 Mon Sep 17 00:00:00 2001
From: Adrian Johnson <ajohnson@redneon.com>
Date: Wed, 28 Sep 2011 11:50:23 +0000
Subject: cairo: fix crash when using poppler_page_get_image()

---
diff --git a/poppler/CairoOutputDev.h b/poppler/CairoOutputDev.h
index effa9dd..0f202a5 100644
--- a/poppler/CairoOutputDev.h
+++ b/poppler/CairoOutputDev.h
@@ -434,6 +434,7 @@ public:
   virtual void stroke(GfxState *state) { }
   virtual void fill(GfxState *state) { }
   virtual void eoFill(GfxState *state) { }
+  virtual void clipToStrokePath(GfxState *state) { }
   virtual GBool tilingPatternFill(GfxState *state, Catalog *cat, Object *str,
 				  double *pmat, int paintType, int tilingType, Dict *resDict,
 				  double *mat, double *bbox,
--
cgit v0.9.0.2-2-gbebe
