https://bugs.gentoo.org/372843

--- mozilla-release/browser/base/Makefile.in
+++ mozilla-release/browser/base/Makefile.in
@@ -62,7 +62,7 @@
 
 include $(topsrcdir)/config/rules.mk
 
-PRE_RELEASE_SUFFIX := ""
+PRE_RELEASE_SUFFIX :=
 
 DEFINES += \
 	-DMOZ_APP_VERSION=$(MOZ_APP_VERSION) \
--- mozilla-release/browser/installer/windows/Makefile.in
+++ mozilla-release/browser/installer/windows/Makefile.in
@@ -51,7 +51,7 @@
 SFX_MODULE = $(topsrcdir)/other-licenses/7zstub/firefox/7zSD.sfx
 APP_VERSION := $(shell cat $(srcdir)/../../config/version.txt)
 DEFINES += -DAPP_VERSION=$(APP_VERSION)
-PRE_RELEASE_SUFFIX := ""
+PRE_RELEASE_SUFFIX :=
 DEFINES += -DPRE_RELEASE_SUFFIX="$(PRE_RELEASE_SUFFIX)"
 
 INSTALLER_FILES = \

