--- glibc-2.15/Makefile
+++ glibc-2.15/Makefile
@@ -200,11 +200,6 @@
 	else $(INSTALL_DATA) $(objpfx)stubs.h $@; fi
 	rm -f $(objpfx)stubs.h
 
-# This makes the Info or DVI file of the documentation from the Texinfo source.
-.PHONY: info dvi pdf html
-info dvi pdf html:
-	$(MAKE) $(PARALLELMFLAGS) -C manual $@
-
 # This makes all the subdirectory targets.
 
 # For each target, make it depend on DIR/target for each subdirectory DIR.
@@ -394,20 +389,9 @@
 	fi
 endif
 
-define format-me
-@rm -f $@
-makeinfo --no-validate --plaintext --no-number-sections $< -o $@
--chmod a-w $@
-endef
-INSTALL: manual/install.texi; $(format-me)
-NOTES: manual/creature.texi; $(format-me)
-manual/dir-add.texi manual/dir-add.info: FORCE
-	$(MAKE) $(PARALLELMFLAGS) -C $(@D) $(@F)
-FAQ: scripts/gen-FAQ.pl FAQ.in
-	$(PERL) $^ > $@.new && rm -f $@ && mv $@.new $@ && chmod a-w $@
 FORCE:
 
-iconvdata/% localedata/% po/% manual/%: FORCE
+iconvdata/% localedata/%: FORCE
 	$(MAKE) $(PARALLELMFLAGS) -C $(@D) $(@F)
 
 # glibc 2.0 contains some header files which aren't used with glibc 2.1
--- glibc-2.15/Makeconfig
+++ glibc-2.15/Makeconfig
@@ -956,8 +956,8 @@
 all-subdirs = csu assert ctype locale intl catgets math setjmp signal	    \
 	      stdlib stdio-common libio malloc string wcsmbs time dirent    \
 	      grp pwd posix io termios resource misc socket sysvipc gmon    \
-	      gnulib iconv iconvdata wctype manual shadow gshadow po argp   \
-	      crypt nss localedata timezone rt conform debug		    \
+	      gnulib iconv iconvdata wctype shadow gshadow argp   \
+	      crypt nss localedata rt conform debug		    \
 	      $(add-on-subdirs) $(dlfcn) $(binfmt-subdir)
 
 ifndef avoid-generated
